### ISSUE TITLE
How to apply aspect to async method

### ISSUE THREAD

**Issue by twzhangyang** [pamidur/aspect-injector#77]
For example I have an async method:
 ```
 [Inject(typeof(ErrorHandlingAspect))]
 public async Task<string> GetSomething()
 {
  //...
  }
```

Then the ErrorHandlingAspect like this:
```  
  [Aspect(Aspect.Scope.Global)]
   public class ErrorHandlingAspect
    {
        [Advice(Advice.Type.Around, Advice.Target.Method)]
        public object HandleMethod([Advice.Argument(Advice.Argument.Source.Name)] string name,
            [Advice.Argument(Advice.Argument.Source.Arguments)] object[] arguments,
            [Advice.Argument(Advice.Argument.Source.Target)] Func<object[], object> method)
        {
            try
                {
                     // ? method is an async method
                     var result = method(arguments);
                     return result
                }
              catch
               {
                    //...
                }
        }
    }
```
My question is does aspect-injector support applying aspect to  async method?

Thanks

**Comment by pamidur** [comment]
Hi, this is a bit complex scenario because:
- you want to handle exception, but you need to return correct result to consumer
- you want to wrap async methods which in fact return Task and the exception might be thrown afterwards

So it seems it requires a bit of magic, like this:
```c#
using AspectInjector.Broker;
using System;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApp2
{
    internal class Program
    {
        private static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            new Program().Do1(true).Wait();
            var d2 = new Program().Do2(true).Result;
            var d3 = new Program().Do3(true).Result;

            new Program().Do1(false).Wait();
            var d21 = new Program().Do2(false).Result;
            var d31 = new Program().Do3(false).Result;
        }

        [Inject(typeof(TestAspect))]
        public async Task Do1(bool error)
        {
            if (error)
                throw new Exception();
        }

        [Inject(typeof(TestAspect))]
        public async Task<int> Do2(bool error)
        {
            if (error)
                throw new Exception();

            return 2;
        }

        [Inject(typeof(TestAspect))]
        public async Task<StringBuilder> Do3(bool error)
        {
            if (error)
                throw new Exception();

            return new StringBuilder();
        }
    }

    [Aspect(Aspect.Scope.Global)]
    internal class TestAspect
    {
        private static MethodInfo _asyncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapAsync), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _syncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapSync), BindingFlags.NonPublic | BindingFlags.Static);

        [Advice(Advice.Type.Around, Advice.Target.Method)]
        public object CatchExceptions(
            [Advice.Argument(Advice.Argument.Source.Target)] Func<object[], object> target,
            [Advice.Argument(Advice.Argument.Source.Arguments)] object[] args,
            [Advice.Argument(Advice.Argument.Source.ReturnType)] Type retType
            )
        {
            if (typeof(Task).IsAssignableFrom(retType))
            {
                var syncResultType = retType.IsConstructedGenericType ? retType.GenericTypeArguments[0] : typeof(object);
                return _asyncErrorHandler.MakeGenericMethod(syncResultType).Invoke(this, new object[] { target, args });
            }
            else
            {
                return _syncErrorHandler.MakeGenericMethod(retType).Invoke(this, new object[] { target, args });
            }
        }


        private static T WrapSync<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return (T)target(args);
            }
            catch (Exception e)
            {
                //log syncEception
                return default(T);
            }
        }

        private static async Task<T> WrapAsync<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return await (Task<T>)target(args);
            }
            catch (Exception e)
            {
                //log asyncEception
                return default(T);
            }
        }
    }
}

```

**Comment by twzhangyang** [comment]
Hi pamidur,

Thanks for your solution, it works. I would say although this scenario is complex but your approach is most simple one comparing other libraries such as 
Castle.DynamicProxy
https://github.com/kswoll/asyncproxy 

Thank you very much. this one is really a simple/flexible approach.

**Comment by saqw3r** [comment]
Hey, @pamidur , it seems I had tried your suggestion, but I've bumped into the known GUI async/await deadlock problem. (I'm using v.1.0.1/1.0.2 of the AspectInjector)

**Comment by saqw3r** [comment]
Generally, the problem is not only about async methods, which returns Task<T>, but also methods with returning types Task and void.

**Comment by pamidur** [comment]
try
```c#
...
return await ((Task<T>)target(args)).ConfigureAwait(false);
...
```

**Comment by saqw3r** [comment]
[Bug with Async.txt](https://github.com/pamidur/aspect-injector/files/3148091/Bug.with.Async.txt)
I've attached an example of MyClass. Here I'm getting an error on line 42:
_asyncTaskExecution.MakeGenericMethod(syncResultType).Invoke(this, new object[] { target, args });

with message: "System.Threading.Tasks.Task WrapAsync(System.Func`2[System.Object[],System.Object], System.Object[]) is not a GenericMethodDefinition. MakeGenericMethod may only be called on a method for which MethodBase.IsGenericMethodDefinition is true." That is reasonable. But then how to call "async Task method"?

**Comment by pamidur** [comment]
this is incorrect approach, in my solution above there are two cases - object and Task<T> methods.
Where any sync method and async void methods are wrapped by object handler; and Task and Task<T> methods are wrapped with Task<T> handler. 
So in fact you don't need separate handlers for Task and Task<T>. Task basically converts to Task<object>.

Though if you want to have a separate handler for non-generic Task methods - you don't need to make this handler generic, I'd suggest something like:
```c#
using AspectInjector.Broker;
using System;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApp2
{
    internal class Program
    {
        private static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            new Program().Do1(true).Wait();
            var d2 = new Program().Do2(true).Result;
            var d3 = new Program().Do3(true).Result;

            new Program().Do1(false).Wait();
            var d21 = new Program().Do2(false).Result;
            var d31 = new Program().Do3(false).Result;
        }

        [TestAspect]
        public async Task Do1(bool error)
        {
            if (error)
                throw new Exception();
        }

        [TestAspect]
        public async Task<int> Do2(bool error)
        {
            if (error)
                throw new Exception();

            return 2;
        }

        [TestAspect]
        public async Task<StringBuilder> Do3(bool error)
        {
            if (error)
                throw new Exception();

            return new StringBuilder();
        }
    }

    [Aspect(Scope.Global)]
    [Injection(typeof(TestAspect))]
    internal class TestAspect : Attribute
    {
        private static MethodInfo _asyncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapAsync), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _asyncGenericErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapAsyncGeneric), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _syncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapSync), BindingFlags.NonPublic | BindingFlags.Static);

        [Advice(Kind.Around)]
        public object CatchExceptions(
            [Argument(Source.Target)] Func<object[], object> target,
            [Argument(Source.Arguments)] object[] args,
            [Argument(Source.ReturnType)] Type retType
            )
        {
            if (typeof(Task).IsAssignableFrom(retType))
            {
                if (retType == typeof(Task))
                    return _asyncErrorHandler.Invoke(this, new object[] { target, args });
                else
                {
                    var syncResultType = retType.IsConstructedGenericType ? retType.GenericTypeArguments[0] : typeof(object);
                    return _asyncGenericErrorHandler.MakeGenericMethod(syncResultType).Invoke(this, new object[] { target, args });
                }
            }
            else
            {
                return _syncErrorHandler.MakeGenericMethod(retType).Invoke(this, new object[] { target, args });
            }
        }


        private static T WrapSync<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return (T)target(args);
            }
            catch (Exception e)
            {
                //log syncEception
                return default(T);
            }
        }

        private static async Task WrapAsync(Func<object[], object> target, object[] args)
        {
            try
            {
                await (Task)target(args);
            }
            catch (Exception e)
            {
                //log asyncEception                
            }
        }

        private static async Task<T> WrapAsyncGeneric<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return await (Task<T>)target(args);
            }
            catch (Exception e)
            {
                //log asyncEception
                return default(T);
            }
        }
    }
}
```

**Comment by saqw3r** [comment]
I've checked your last approach, but it generates the Exception 
at the line:  `await (Task)target(args);`
at the method WrapAsync(...) with next details:
```
Message: The type 'System.Void' may not be used as a type argument.

StackTrace:
 at System.RuntimeType.ThrowIfTypeNeverValidGenericArgument(RuntimeType type)
   at System.RuntimeType.SanityCheckGenericArguments(RuntimeType[] genericArguments, RuntimeType[] genericParamters)
   at System.Reflection.RuntimeMethodInfo.MakeGenericMethod(Type[] methodInstantiation)
   at MyClass.Handle(Func`2 target, Object[] args, MethodBase method, String methodName, Type returnType) in D:\MyApp\MyApp
```

**Comment by saqw3r** [comment]
[ConsoleApp1.zip](https://github.com/pamidur/aspect-injector/files/3149027/ConsoleApp1.zip)
Attached an adopted example for .net 4.5.2

**Comment by pamidur** [comment]
In earlier version injection were done directly, so your code is slighly incorrect. I fixed it for you and also added handlers for void methods. Note that you cannot really catch Exceptions in ```async void``` methods, and it has nothing to do with AspectInjector.


```c#
using AspectInjector.Broker;
using System;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using AspectInjector;
using static AspectInjector.Broker.Aspect;
using static AspectInjector.Broker.Advice;
using static AspectInjector.Broker.Advice.Argument;

namespace ConsoleApp2
{
    internal class Program
    {
        private static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            new Program().Do1(true).Wait();
            var d2 = new Program().Do2(true).Result;
            var d3 = new Program().Do3(true).Result;
            new Program().Do4(true);
            var d5 = new Program().Do5(true);
            var d6 = new Program().Do6(true);
            new Program().Do7(true);


            new Program().Do1(false).Wait();
            var d21 = new Program().Do2(false).Result;
            var d31 = new Program().Do3(false).Result;
            new Program().Do4(false);
            var d51 = new Program().Do5(false);
            var d61 = new Program().Do6(false);
            new Program().Do7(false);
        }



        [Inject(typeof(TestAspect))]
        public async Task Do1(bool error)
        {
            if (error)
                throw new Exception();
        }

        [Inject(typeof(TestAspect))]
        public async Task<int> Do2(bool error)
        {
            if (error)
                throw new Exception();

            return 2;
        }

        [Inject(typeof(TestAspect))]
        public async Task<StringBuilder> Do3(bool error)
        {
            if (error)
                throw new Exception();

            return new StringBuilder();
        }

        [Inject(typeof(TestAspect))]
        public async void Do4(bool error)
        {
            if (error)
                throw new Exception();
        }

        [Inject(typeof(TestAspect))]
        public object Do5(bool error)
        {
            if (error)
                throw new Exception();

            return new StringBuilder();
        }

        [Inject(typeof(TestAspect))]
        public int Do6(bool error)
        {
            if (error)
                throw new Exception();

            return 1;
        }

        [Inject(typeof(TestAspect))]
        public void Do7(bool error)
        {
            if (error)
                throw new Exception();
        }
    }

    [Aspect(Scope.PerInstance)]
    internal class TestAspect
    {
        private static MethodInfo _asyncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapAsync), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _asyncGenericErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapAsyncGeneric), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _syncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapSync), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _syncGenericErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapSyncGeneric), BindingFlags.NonPublic | BindingFlags.Static);

        [Advice(Advice.Type.Around, Target.Method)]
        public object CatchExceptions(
            [Argument(Source.Target)] Func<object[], object> target,
            [Argument(Source.Arguments)] object[] args,
            [Argument(Source.ReturnType)] System.Type retType
            )
        {
            if (typeof(Task).IsAssignableFrom(retType))
            {
                if (retType == typeof(Task))
                    return _asyncErrorHandler.Invoke(this, new object[] { target, args });
                else
                {
                    var syncResultType = retType.IsConstructedGenericType ? retType.GenericTypeArguments[0] : typeof(object);
                    return _asyncGenericErrorHandler.MakeGenericMethod(syncResultType).Invoke(this, new object[] { target, args });
                }
            }
            else
            {
                if (retType == typeof(void))
                    return _syncErrorHandler.Invoke(this, new object[] { target, args });
                else
                    return _syncGenericErrorHandler.MakeGenericMethod(retType).Invoke(this, new object[] { target, args });
            }
        }

        private static void WrapSync(Func<object[], object> target, object[] args)
        {
            try
            {
                target(args);
            }
            catch (Exception e)
            {
                //log syncEception
            }
        }
        private static T WrapSyncGeneric<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return (T)target(args);
            }
            catch (Exception e)
            {
                //log syncEception
                return default(T);
            }
        }

        private static async Task WrapAsync(Func<object[], object> target, object[] args)
        {
            try
            {
                await (Task)target(args);
            }
            catch (Exception e)
            {
                //log asyncEception                
            }
        }

        private static async Task<T> WrapAsyncGeneric<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return await (Task<T>)target(args);
            }
            catch (Exception e)
            {
                //log asyncEception
                return default(T);
            }
        }
    }
}
```

**Comment by saqw3r** [comment]
[WebApplication1.zip](https://github.com/pamidur/aspect-injector/files/3152700/WebApplication1.zip)
It appears, that the issue is not reproducible in Console application, but reproducible in Web application (probably, it's somehow related to the fact, that in a console application, there is no synchronization contex). 
I created out of the box ASP.NET Web Mvc application. And added pretty similar (but a bit different) code in Contact action. And there is the deadlock. If remove attribute from AsyncTestClass the deadlock is passing away.

**Comment by saqw3r** [comment]
@twzhangyang Why? It's still unresolved issue for me.

**Comment by twzhangyang** [comment]
Hi @saqw3r 
Sorry man, Did not notice your question is still in there. but for this topic it already worked for me and I forgot to close it.
I guess you have another issue and different with original question. Could you please open another ticket?

**Comment by saqw3r** [comment]
@twzhangyang No, you closed it. And then I reopened. You can unsubscribe in case you don't interested in our conversation and notification is bothering you. 
I have the issue, which is directly related to the topic. But in case, author of framework would decide that it worths another ticket to be open - I will do it.

**Comment by pamidur** [comment]
Guys, I guess the question is related, because definitely in some cases we might capture the context, thus causing deadlock.

There is a slight chance however that this issue is only present in the old version. In this case we'll move to another issue for legacy support.

I need a time to figure that out. Until then I'll leave open.

**Comment by pamidur** [comment]
@saqw3r I've added configure await to wrappers and it works fine for me:
<details>
  <summary>Code</summary>

```c#
using System;
using System.Web.Mvc;
using AspectInjector.Broker;
using System.Reflection;
using System.Threading.Tasks;
using static AspectInjector.Broker.Aspect;
using static AspectInjector.Broker.Advice;
using static AspectInjector.Broker.Advice.Argument;

namespace WebApplication1.Controllers
{
    public class HomeController : Controller
    {
        public ActionResult Index()
        {
            return View();
        }

        public ActionResult About()
        {
            ViewBag.Message = "Your application description page.";

            return View();
        }

        public ActionResult Contact()
        {
            ViewBag.Message = "Your contact page.";

            AsyncTestClass asyncTest = new AsyncTestClass();

            var task = asyncTest.MakeSomeTaskAsynchronously();
            Task.WaitAll(new Task[] { task });

            return View();
        }
    }

    [Aspect(Scope.PerInstance)]
    internal class TestAspect
    {
        private static MethodInfo _asyncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapAsync), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _asyncGenericErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapAsyncGeneric), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _syncErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapSync), BindingFlags.NonPublic | BindingFlags.Static);
        private static MethodInfo _syncGenericErrorHandler = typeof(TestAspect).GetMethod(nameof(TestAspect.WrapSyncGeneric), BindingFlags.NonPublic | BindingFlags.Static);

        [Advice(Advice.Type.Around, Target.Method)]
        public object Handle(
            [Argument(Source.Target)] Func<object[], object> target,
            [Argument(Source.Arguments)] object[] args,
            [Argument(Source.ReturnType)] System.Type retType
            )
        {
            if (typeof(Task).IsAssignableFrom(retType))
            {
                if (retType == typeof(Task))
                    return _asyncErrorHandler.Invoke(this, new object[] { target, args });
                else
                {
                    var syncResultType = retType.IsConstructedGenericType ? retType.GenericTypeArguments[0] : typeof(object);
                    return _asyncGenericErrorHandler.MakeGenericMethod(syncResultType).Invoke(this, new object[] { target, args });
                }
            }
            else
            {
                if (retType == typeof(void))
                    return _syncErrorHandler.Invoke(this, new object[] { target, args });
                else
                    return _syncGenericErrorHandler.MakeGenericMethod(retType).Invoke(this, new object[] { target, args });
            }
        }

        private static void WrapSync(Func<object[], object> target, object[] args)
        {
            try
            {
                target(args);
            }
            catch (Exception e)
            {
                int i = 1;
                //log syncEception
                return;
            }
        }
        private static T WrapSyncGeneric<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return (T)target(args);
            }
            catch (Exception e)
            {
                int i = 1;
                //log syncEception
                return default(T);
            }
        }

        private static async Task WrapAsync(Func<object[], object> target, object[] args)
        {
            try
            {
                await ((Task)target(args)).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                int i = 1;

                //log asyncEception    

                return;
            }
        }

        private static async Task<T> WrapAsyncGeneric<T>(Func<object[], object> target, object[] args)
        {
            try
            {
                return await ((Task<T>)target(args)).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                int i = 1;
                //log asyncEception
                return default(T);
            }
        }
    }

    [Inject(typeof(TestAspect))]
    public class AsyncTestClass
    {
        public async Task MakeSomeTaskAsynchronously()
        {
            await Task.Delay(1000).ConfigureAwait(false);
        }
    }
}
```
</details>

**Comment by pamidur** [comment]
@saqw3r please let us know if the above solution works for you

**Comment by pamidur** [comment]
I've added updated sample to samples subdir here
https://github.com/pamidur/aspect-injector/tree/master/samples/UniversalWrapper

Now it handles all possible scenarios without capturing context (necessary for old asp.net frameworks).
The only thing missing is `async void` exception handling, it requires execution via custom TaskFactory, which is a bit tricky as we might break something in the execution flow. I'll looking into it later. Until them, if no other question, I'll close this issue.

**Comment by marektoman** [comment]
@pamidur I wonder is there a reason why the non-generic methods are invoked via MethodInfo in your example?

**Comment by pamidur** [comment]
@marektoman , no reason, I've just copy-pasted :)
However in the sample there are generic methods only
https://github.com/pamidur/aspect-injector/tree/master/samples/UniversalWrapper

**Comment by matiaspecile** [comment]
@pamidur I realize it's 5 years after the last comment on this, but I think the library is essentially in the same state relating to this issue (if not, please let me know!)

so given that, along the lines of marktoman's question, why is there any special care taken for synchronous methods?  I would have expected non-async stuff to work out of the box without any extra care taken for its return type.  is there something I'm missing in my understanding?

**Comment by pamidur** [comment]
Hi @matiaspecile , if you mean the substitution of return void to return object - the reason is twofold:
- aspect's advice has a signature with the object is returned for Around aspects. Because there is no way to know in advance where are you going to apply said aspect. Returned value is ignored in case the method is actually void 
- Universal aspect library is there to hide this detail. This its approach is to generate Expression with above mentioned signature from one side, but with ability to handle all types of methods from the other.

Please let me know if I got your question right or if you need more details.

**Comment by matiaspecile** [comment]
@pamidur thank you very much for the quick look and information.  I'm still a little unclear, so let me add a bit more info to my question

I'm referencing your example above because it's a little less dynamic than the universal aspect example (I think, in practice, both cases are the same though)

```
private static T WrapSyncGeneric<T>(Func<object[], object> target, object[] args)
{
    try
    {
        return (T)target(args);
    }
    catch (Exception e)
    {
        int i = 1;
        //log syncEception
        return default(T);
    }
}
```

what is the purpose of wrapping the target `Func` here?  the return type is casted to its actual type to be casted back to object in the `Around` method.  is the whole point of this to provide a default value for the actual type in the event an exception is caught?

(the async wrappers make a lot more sense because it's obvious they're there to await the `Task` with either a return type - `Task<T>` - or no return type - `Task`)

> This its approach is to generate Expression with above mentioned signature from one side, but with ability to handle all types of methods from the other.

to sort of add to what I'm trying to understand - given your above comment, I thought the library already handled synchronous methods with either void or any return type when it returned `object` in the `Around` method

**Comment by pamidur** [comment]
Hi @matiaspecile , yes I had another look at this code. Here yes indeed the purpose was to have default value in case of exception. This is what OP of this issue wanted. 
There are ways to simplify this, there is no need to cast it back and forth.
Universal aspect doesn't return default value though. 
Tldr. It can be simplified
